File: algorithm/algorithm.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Macros */
#define ABS(x) ((x) < 0 ? -(x) : (x))
#define WINDOW(R) (2 * (R) + 1)

/* Data-related constants */
#define DATA_FILE   "scope.csv"
#define OUTPUT_FILE "output.csv"
#define FS          360

/* QRS algorithm parameters */
#define N           25  // Moving average window size for high-pass filter
#define S           7   // Triangle template matching parameter
#define L           5   // Moving average window size for Low-pass filter
#define M           150 // Moving average window size for dynamic threshold calculation
#define BETA        2.5 // Dynamic threshold coefficient
#define MIN_RR_DIST (0.272 * FS) // Minimum distance between two RR peaks in seconds

/* Buffer parameters */
#define NOTCHED_SIZE (WINDOW(N) + 1)
#define H_SIZE (FS * 2)
#define T_SIZE (WINDOW(L) + 1)
#define L1_SIZE (WINDOW(M) + 1)

/* Notch filter coefficients */
static float B[] = {0.9175, -0.9451, 0.9175};
static float A[] = {0, -0.9451, 0.8350};

unsigned short max_index(unsigned short arr[],
                         unsigned short start_idx,
                         unsigned short end_idx) {

    unsigned short max_value = arr[start_idx];
    unsigned short max_index = start_idx;

    for (unsigned short i = start_idx + 1; i <= end_idx; i++) {
        if (arr[i] > max_value) {
            max_value = arr[i];
            max_index = i;
        }
    }

    return max_index;
}

int rounded(float value) {
    if (value >= 0) {
        return (int)(value + 0.5f);
    } else {
        return (int)(value - 0.5f);
    }
}

int main() {
    // Open the data and output files
    FILE *input_file = fopen(DATA_FILE, "r");
    FILE *output_file = fopen(OUTPUT_FILE, "w");

    // Signal buffers
    unsigned short x[3];
    float y[3];
    unsigned short notched[NOTCHED_SIZE];
    unsigned short h[H_SIZE];
    float t[T_SIZE];
    float l1[L1_SIZE];

    // Buffer initializations
    memset(x, 0, sizeof(x));
    memset(y, 0, sizeof(y));

    // Signal values
    unsigned short input;
    float y_val;
    unsigned short notched_val;
    float notched_bar_val = 0;
    short h_hat_val = 0;
    unsigned short h_val = 0;
    float t_val = 0;
    float l1_val = 0;
    float l2_val = 0;
    float th_val = 0;
    float theta;

    // Buffer indices
    unsigned short i;
    unsigned short j;
    unsigned short i_notched = 0;
    unsigned short i_h = 0;
    unsigned short i_h_hat = 0;
    unsigned short i_ecg = 0;
    unsigned short i_t = 0;
    unsigned short i_l1 = 0;

    // Area of interest flags
    unsigned char prev_aoi;
    unsigned char aoi = 0;
    
    // Peak detection indices
    unsigned short i_onset;
    unsigned short i_offset;
    unsigned short i_cand_max;
    unsigned short i_curr_max = 0;
    unsigned short i_prev_max;

    float rr; // Distance (in samples) between most recent two RR peaks
    float bpm = 0; // beats per minute reading
    
    // Read data from file and store in buffer, then write to output file
    while (fscanf(input_file, "%i,%*s", &input) != EOF) {

        for (i = 0; i < 2; i++) {
            x[i] = x[i + 1];
            y[i] = y[i + 1];
        }
        x[2] = input;
    
        y_val = 0;
        for (i = 0; i < 3; i++) {
            y_val += B[i] * x[2 - i];
            y_val -= A[i] * y[2 - i];
        }
        y[2] = y_val;

        notched_val = (unsigned short) y_val;
        if (i_notched == NOTCHED_SIZE) {
            for (i = 0; i < NOTCHED_SIZE - 1; i++) {
                notched[i] = notched[i + 1];
            }
            i_notched--;
        }
        notched[i_notched++] = notched_val;

        notched_bar_val += (float) notched_val / WINDOW(N);
        if (i_notched > WINDOW(N)) {
            notched_bar_val -= (float) x[0] / WINDOW(N);

            h_hat_val = notched_val - notched_bar_val;

            h_val = ABS(h_hat_val);
            if (i_h == H_SIZE) {
                for (i = 0; i < H_SIZE - 1; i++) {
                    h[i] = h[i + 1];
                }
                i_h--;
                i_onset--;
                i_curr_max--;
                i_prev_max--;
            }
            h[i_h++] = h_val;

            if (i_h >= WINDOW(S)) {
                j = i_h - 1;
                t_val = (h[j - S] - h[j - (2 * S)]) * (h[j - S] - h[j]);

                if (i_t == T_SIZE) {
                    for (i = 0; i < T_SIZE - 1; i++) {
                        t[i] = t[i + 1];
                    }
                    i_t--;
                }
                t[i_t++] = t_val;

                l1_val += t_val / WINDOW(L);
                if (i_t > WINDOW(L)) {
                    l1_val -= t[0] / WINDOW(L);

                    if (i_l1 == L1_SIZE) {
                        for (i = 0; i < L1_SIZE - 1; i++) {
                            l1[i] = l1[i + 1];
                        }
                        i_l1--;
                    }
                    l1[i_l1++] = l1_val;

                    l2_val += l1_val / WINDOW(M);
                    if (i_l1 > WINDOW(M)) {
                        l2_val -= l1[0] / WINDOW(M);
                    }
                    if (i_l1 > M + 1) {
                        theta = 0.25 * l2_val;
                        th_val = BETA*l2_val + theta;

                        prev_aoi = aoi;
                        aoi = l1_val >= th_val ? 1 : 0;

                        if (aoi - prev_aoi == 1) {
                            i_onset = i_h;
                        } else if (aoi - prev_aoi == -1) {
                            i_offset = i_h;
                            i_cand_max = max_index(h, i_onset, i_offset);

                            if (i_curr_max == 0) {
                                i_curr_max = i_cand_max;
                            } else {
                                if (i_cand_max - i_curr_max < MIN_RR_DIST) {
                                    if (h[i_cand_max] > h[i_curr_max]) {
                                        i_curr_max = i_cand_max;
                                    } 
                                } else {
                                    i_prev_max = i_curr_max;
                                    i_curr_max = i_cand_max;
                                    rr = i_curr_max - i_prev_max;
                                    bpm = 60.0 * FS / (float) rr;
                                    printf("bpm: %f\n", bpm);
                                }
                            }
                        }           
                    }
                }
            }
        }
        fprintf(output_file, "%i,%i,%f,%f,%f,%i\n", notched_val, h_val, t_val, l1_val, th_val, aoi);
    }
    fclose(input_file);
    fclose(output_file);

    return 0;
}

File: algorithm/algorithm.m
% Author: Steve Farra
%
% Description: This Octave script is used to prototype the software driving the
%              ec-stingy, a low-cost, fully open-sourced electrocardiogram.
%              the script has two components:
%
%              - Notch filter design and implementation, used to filter out
%              power line interference so that the signal may be used as a
%              cardiac monitor.'
%
%              - An end-to-end R-peak detection algorithm implementation for
%              the heart rate monitor. This algorithm closely follows a 2019
%              paper, "Low Resource Complexity R-peak Detection Based on
%              Triangle Template Matching and Moving Average Filter" by
%              Tam Nguyen, Xiaoli Qin, Anh Dinh, and Francis Bui. More details
%              can be found in the README.
%
%              Heart rate readings are printed to the console, in addition to
%              several plots that are displayed showing the filter(s) used by
%              the cardiac and heart rate monitors.
%
% License: This script is licensed under the MIT license.
%          See the accompanying LICENSE file for the full text of the license.

% Used for the FFT and notch filter sections of the script.
pkg load signal;

% The moving average window lengths in the R-peak detection paper used
% specified as one-sided. This macro is used to convert between the two.
window = @(R) 2 * R + 1;

% Sampling frequency. The paper uses the typical rate of 360 Hz, commonly used
% in databases such as the MIT-BIH Arrhythmia Database. The parameters which
% follow are dependent on the sampling frequency, so if you change it be sure
% to adjust the other parameters according to the paper.
Fs = 360;

% One-sided window length for the high-pass filter, used to remove DC noise
% and other artefacts such as baseline wander.
N = 25;

% Parameter for the triangle template matching filter.
s = 7;

% One-sided window length for the first low-pass filter. The output of this
% filter is used to smoothen out the R-peaks.
L = 5;

% One-sided window length for the second low-pass filter. The output of this
% filter is used to set the threshold value.
M = 150;

# Coefficient used to scale up the threshold value.
beta = 2.5;

% Since there may be several detected R-peaks within an area of interest, the
% paper uses the assumption that the maximum heart rate is 206 bpm, or a
% minimum of 272 ms between R-peaks, as error correction.
MIN_RR_DIST_MS = 272;

% File containing the ECG signal used to test R-peak detection. Modify the
% filename to your liking.
data_file = 'scope.csv';
data = csvread(data_file);
x = data(:,1);

% Before carrying out the R-peak detection algorithm, we do a preliminary
% analysis of the frequency content of our signal. In our case, there is a high
% amount of powerline interference present at 59 Hz since our electrode cables
% don't have any shielding.
signal_length = length(x);
frequency_resolution = Fs / signal_length;
frequency_axis = (0:(signal_length/2)) * frequency_resolution;
signal_no_dc = x - mean(x);
signal_fft = fft(signal_no_dc);
magnitude_spectrum = abs(signal_fft);
magnitude_spectrum = magnitude_spectrum(1:(signal_length/2 + 1));

% We use a Notch filter to eliminate powerline interference. Be sure to adjust
% the frequency and bandwidth of the filter according to your needs. Keep in
% mind this filter is not included in the paper and is unnecessary for R-peak
% detection. However, we include it so that our ECG can double as a cardiac
% monitor in addition to being a heart rate monitor.
notch_frequency = 59;
notch_bandwidth = 5;
notch_frequency_normalized = notch_frequency / (Fs/2);
notch_bandwidth_normalized = notch_bandwidth / (Fs/2);
[b, a] = pei_tseng_notch(notch_frequency_normalized,
                         notch_bandwidth_normalized);
notched = filter(b, a, x);

% High-pass filter, implemented as stated in the paper.
notched_bar = movmean(notched, window(N));
h_hat = notched - notched_bar;
h = abs(h_hat);

% Triangle template matching filter, implemented as stated in the paper.
t = zeros(length(h), 1);
for i = s+1:length(h)-s
    t(i) = (h(i) - h(i-s)) * (h(i) - h(i+s));
end

% First low-pass filter, implemented as stated in the paper.
l = movmean(t, window(L));

% Threshold calculation.  The paper suggests using the statistical mean of the
% output of the second low-pass filter as the value of theta.
ma = movmean(l, window(M));
theta = mean(l)/4;
th = beta*ma + theta;

% Holds current peak candidate values.
buffer = [];

% Indicates whether a peak candidate is under consideration.
flag = 0;

% Array to hold the detected R-peak indices.
r_peaks = [];

% Index of the previously detected R-peak, initialized to -1 to denote
% that a peak hasn't been detected yet.
prev_r_peak_idx = -1;

for i = 1:length(l)
    % Check if the current value is above the threshold.
    if l(i) >= th(i)
        % If no peak is under consideration, start considering the current one.
        if flag == 0
            % Reset the buffer for a new peak.
            buffer = [];

            % Set the flag to indicate a peak is under consideration.
            flag = 1;
        end
        % Append the current value to the buffer since the signal has met the
        % threshold. The output of the high pass filter is used for this.
        buffer = [buffer; i, h(i)];
  % If the current value is below threshold,
  else
        % and a peak was just previously under consideration,
        if flag == 1
            % find the maximum value in the buffer. This is a candidate R-peak.
            [~, max_idx] = max(buffer(:, 2));

            % Get the index of the maximum value, which is what we're interested
            % in.
            r_peak_idx = buffer(max_idx, 1);

            % If there was a previously detected R-peak,
            if prev_r_peak_idx ~= -1

                % we perform error correction by computing the distance to it.
                r_peak_distance = (r_peak_idx - prev_r_peak_idx) / Fs * 1000;

                % If the distance is too short,
                if r_peak_distance <= MIN_RR_DIST_MS

                    % and the current R-peak is higher than the previous one,
                    if h(r_peak_idx) >= h(prev_r_peak_idx)

                        % discard the previous R-peak.
                        r_peaks(end) = [];

                    % Otherwise,
                     else

                        % discard the current one.
                        continue;
                    end
                end
            end

            % Add the current R-peak to the list of detected R-peaks.
            r_peaks = [r_peaks; r_peak_idx];

            % Update the index of the last detected R-peak.
            prev_r_peak_idx = r_peak_idx;
        end

        % Reset the flag to indicate no peak is under consideration.
        flag = 0;
    end
end

% Heart rate readings displayed in units of bpm.
r_peak_times = r_peaks / Fs;
bpm_readings = 60 ./ diff(r_peak_times);
disp('bpm readings:');
disp(bpm_readings);

% x-axis for all the plots that follow in this script.
t_plot = (1:length(x)) / Fs;

figure(1);

% Plot of the ECG signal imported by the user.
subplot(3,1,1);
plot(t_plot, x);
xlabel('Time (s)');
ylabel('Amplitude');
title('Original ECG Signal');

% Frequency spectrum of the ECG signal.
subplot(3,1,2);
plot(frequency_axis, magnitude_spectrum);
xlabel('Frequency (Hz)');
ylabel('Magnitude');
title('Magnitude Spectrum (DC noise removed)');

% The original ECG signal after the notch filter has been applied to it.
subplot(3,1,3);
plot(t_plot, notched);
xlabel('Time (s)');
ylabel('Amplitude');
title('ECG Signal (Powerline Interference Removed)');

figure(2);

% Output of the high-pass filter with markers to indicate where R-peaks have
% been detected.
subplot(3,1,1);
plot(t_plot, h);
hold on;
scatter(r_peak_times, h(r_peaks), 'r', 'filled');
hold off;
xlabel('Time (s)');
ylabel('Amplitude');
title('High-pass Filtered ECG Signal with R-peaks Overlay');
legend('Filtered ECG', 'R-peaks');

% Output of the triangle template matching filter.
subplot(3,1,2);
plot(t_plot, t);
xlabel('Time (s)');
ylabel('Amplitude');
title('Triangle Template Matching Output');

% Output of the first low-pass filter with the threshold value superimposed.
% wherever the former is above the latter is where the R-peak search is
% performed.
subplot(3,1,3);
plot(t_plot, l, 'b');
hold on;
plot(t_plot, th, 'r');
hold off;
xlabel('Time (s)');
ylabel('Amplitude');
title("Low-pass Filtered Triangle Template Matching Output \
with Modified Threshold Overlay");
legend('Low-pass Filtered', 'Modified Threshold');

